src/emit/expression.rs

```diff
@@ -2121,3 +2123,227 @@ fn runtime_cast<'a>(
         _ => unreachable!(),
     }
 }
+
+/// Emit expression into a slice
+pub(super) fn expression_to_slice<'a, T: TargetRuntime<'a> + ?Sized>(
+    target: &T,
+    bin: &Binary<'a>,
+    e: &Expression,
+    to: &Type,
+    vartab: &HashMap<usize, Variable<'a>>,
+    function: FunctionValue<'a>,
+    ns: &Namespace,
+) -> (PointerValue<'a>, IntValue<'a>) {
+    emit_context!(bin);
+
+    let Type::Slice(to_elem_ty) = to else {
+        unreachable!()
+    };
+
+    let llvm_to = bin.llvm_type(to, ns);
+
+    match e {
+        Expression::ArrayLiteral {
+            dimensions, values, ..
+        } => {
+            let length = dimensions[0];
+
+            let llvm_length = i32_const!(length.into());
+
+            let output = bin.build_array_alloca(function, llvm_to, llvm_length, "seeds");
+
+            for i in 0..length {
+                let (ptr, len) = expression_to_slice(
+                    target,
+                    bin,
+                    &values[i as usize],
+                    to_elem_ty,
+                    vartab,
+                    function,
+                    ns,
+                );
+
+                let output_ptr = unsafe {
+                    bin.builder.build_gep(
+                        llvm_to,
+                        output,
+                        &[i32_const!(i.into()), i32_zero!()],
+                        "output_ptr",
+                    )
+                };
+
+                bin.builder.build_store(output_ptr, ptr);
+
+                let output_len = unsafe {
+                    bin.builder.build_gep(
+                        llvm_to,
+                        output,
+                        &[i32_const!(i.into()), i32_const!(1)],
+                        "output_len",
+                    )
+                };
+
+                bin.builder.build_store(output_len, len);
+            }
+
+            return (output, llvm_length);
+        }
+        Expression::AllocDynamicBytes {
+            initializer: Some(initializer),
+            ..
+        } => {
+            let ptr = bin.emit_global_string("slice_constant", initializer, true);
+            let len = i64_const!(initializer.len() as u64);
+
+            return (ptr, len);
+        }
+        _ => (),
```

Comment:
This is a bit confusing, I expected an `unreachable!()` instead. Is this enforced in an earlier stage? Maybe a comment explaining why this is fine here would be nice

Reply:
You're right, the logic is strange here. I've re-written it, let me know what you think.

Reply:
Ah I see. Yes that's much shorter and quicker to grasp.

However, I think I overlooked the early return in case of `ArrayLength::Dynamic`, that might actually play a role here?

Reply:
Easy to grasp now, thanks!